keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+  params: "replica-limit-nontest.example.com"
+  match:
+    namespaceSelectors:
+    - key: environment,
+      operator: Exists
+  enforcement: [Deny]
+```
+
+With this binding, the test and global policy bindings overlap. Resources
+admitted to test environment would then be checked against both policy
+configurations.
+
+To summarize, this "API Shape" separates admission policy _definition_ from
+_configuration_. This has a couple advantages:
+
+- Access to, and delegation of, policy configuration is more manageable. In
+  particular, Kubernetes RBAC works well with this design.
+- Without the separation, the next most obvious API shape would be to encode
+  everything into a single resource, which could easily become very large and
+  run into resource size limits.
+
+##### Policy Definitions
+
+Policy definitions are responsible for:
+
+- Defining what validations the policy enforces and how violations are reported 
+- Defining how a policy may be configured
+
+Each `ValidatingAdmissionPolicy` resource contains a `spec.match` to declare
+what resources it validates. This field is required.
+
+- `spec.match` constrains which resources this policy can be applied to. Policy
+  bindings each have match rules with further narrow this constraint, but cannot
+  expand it. This allows the CEL expressions to make safe assumptions. E.g. a
+  CEL expression that is constrained to CREATES and UPDATES of resources is
+  guaranteed the root `object` variable is never null, but a CEL expression that
+  might need to evaluate a DELETE must handle the root `object` variable being
+  null. See below "Match criteria" section for how match criteria is described.
+- `spec.match` guides CEL expression type checking, see the below "Type safety"
+  section for more details. 
+
+CEL expressions have access to the contents of the `AdmissionReview` type,
+organized into CEL variables as well as some other useful variables:
+
+- 'object'
+- 'oldObject'
+- 'review'
+  - 'requestResource' (GVR)
+  - 'resource' (GVR)
+  - 'name'
+  - 'namespace'
+  - 'operation'
+  - 'userInfo'
+  - 'dryRun'
+  - 'options'
+- 'config' - configuration data of the policy configuration being validated
+
+See below "Reporting violations to Clients" for more detail about how the
+`spec.validations` field works and how violations are reported.
+
+##### Policy Configuration
+
+`PolicyBinding` resources and parameter CRDs together define how cluster
+administrators configure policies for clusters.
+
+Each `PolicyBinding` contains:
+
+- `spec.policy` - A reference to the policy being configured
+- `spec.match` - Match criteria for which resources the policy should validate
+- `spec.params` - Reference to the custom resource containing the params to use
+  when validating resources 
+- `spec.enforcement` - How validation failures should be enforced
+
+Example:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: PolicyBinding
+metadata:
+  name: "xyzlimit-scale.example.com"
```

Comment:
This name can't work since there can be multiple.

Reply:
Oops, fixing the examples.